packages/database/src/migrations/20210121213045-MigrateStriveCanonicalToAlternateHeirarchy-modifies-data.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+import { arrayToDbString, generateId, insertObject, updateValues } from '../utilities';
+
+var dbm;
+var type;
+var seed;
+
+/**
+ * We receive the dbmigrate dependency from dbmigrate initially.
+ * This enables us to not have to rely on NODE_PATH.
+ */
+exports.setup = function (options, seedLink) {
+  dbm = options.dbmigrate;
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+const striveHierarchyId = '5e9d06e261f76a30c4000018';
+const countryCode = 'PG';
+const countryId = '5d3f8844f327a531bfd8ad77';
+const canonicalTypes = '{country,district,facility,village,case}';
+
+// Explicity naming codes so not to be dependant on order of operations of importing districts for FEPT
+const striveDistrictCodes = [
+  'PG_Sandaun',
+  'PG_National Capital District',
+  'PG_Milne Bay ',
+  'PG_New Ireland',
+  'PG_Morobe',
+  'PG_Simbu',
+  'PG_Western',
+  'PG_Madang',
+];
+
+const selectDistrictIds = (db, districtCodes) =>
+  db.runSql(`select id from "entity" where code in (${arrayToDbString(districtCodes)})`);
+
+const addDistrictEntityRelations = (db, ids) =>
+  ids.map(id =>
+    insertObject(db, 'entity_relation', {
+      id: generateId(),
+      parent_id: countryId,
+      child_id: id.id,
+      entity_hierarchy_id: striveHierarchyId,
+    }),
+  );
+
+const selectEntities = (db, entityType) =>
+  db.runSql(
+    `select id, parent_id from "entity" where country_code = '${countryCode}' and type = '${entityType}'`,
+  );
+
+const addEntityRelations = (db, entities) =>
+  entities.map(entity =>
+    insertObject(db, 'entity_relation', {
+      id: generateId(),
+      parent_id: entity.parent_id,
+      child_id: entity.id,
+      entity_hierarchy_id: striveHierarchyId,
+    }),
+  );
+
+exports.up = async function (db) {
+  const districtIds = (await selectDistrictIds(db, striveDistrictCodes)).rows;
+  await addDistrictEntityRelations(db, districtIds);
+
+  const facilities = (await selectEntities(db, 'facility')).rows;
+  await addEntityRelations(db, facilities);
+
+  const villages = (await selectEntities(db, 'village')).rows;
+  await addEntityRelations(db, villages);
+
+  const cases = (await selectEntities(db, 'case')).rows;
+  await addEntityRelations(db, cases);
+
+  await updateValues(db, 'entity_hierarchy', { canonical_types: {} }, { id: striveHierarchyId });
```

Comment:
@edmofro is this correct - we ewant to remove canonical types from Strive when creating as alternate_heirarchy?

Reply:
I probably need to understand the context a bit more, but my initial reaction is "no".

`canonical_types` indicates which types should be automatically included in the hierarchy. If you remove that, how will new cases be linked in?

Further info: even if you add `case` into the `canonical_types`, by adding the cases explicitly as relations, you're telling the hierarchy traverser that this level in the hierarchy has been manually taken care of, so it won't bother looking to include any automatically.

Why is this being done?

Happy to have a call on this if you want?

Reply:
In summary from call:
All hierarchies contain a mixture of 'canonical' and 'alternate' relations. `entity_relation` entries entered on a type 'generational' basis where divergent from canonical rather than an all or nothing.
`canonical_types` should remain for Strive, removing `sub_district` which do not exist for Strive.
`entity_relation` entries should be added for the existing 8 districts county->district entities. The rest will continue to function as before.
When importing districts (Provinces) and sub_districts (Districts) for FETP project [#1671](https://github.com/beyondessential/tupaia-backlog/issues/1671) `canonical_types` for FETP hierarchy will be 'country,district,sub_district,individual` and project->country relation defined as `entity_relation`